required_fields.node.js

```diff
@@ -19,29 +26,618 @@ function castBool(value) {
+        const {heating, cooling} = system;
+        heating && hvac_types.push(heating.type);
+        cooling && hvac_types.push(cooling.type);
+    });
+    if(!hvac_types.includes('boiler') && category === 'combined') {
+        addErrorMessage(errorMessages[ENUMS.ERROR], `building/systems/domestic_hot_water/category`, 'Must have a boiler for combined hot water category');
+    }
+}
+
+function checkHotWaterFuelValid(hot_water, errorMessages) {
+    const {type, fuel_primary} = hot_water;
+    if(['tankless_coil', 'indirect'].includes(type) && !nullOrUndefined.includes(fuel_primary)) {
+        addErrorMessage(errorMessages[ENUMS.ERROR], `building/systems/domestic_hot_water/fuel_primary`, 'Fuel is only used if type is set to storage or heat pump');
+    } else if(type === 'heat_pump' && fuel_primary !== 'electric') {
+        addErrorMessage(errorMessages[ENUMS.ERROR], `building/systems/domestic_hot_water/fuel_primary`, 'Fuel must be electric if type is heat pump');
+    }
+}
+
+function checkHotWaterEfficiencyValid(hot_water, errorMessages) {
+    const {type, efficiency_method} = hot_water;
+    if(['heat_pump', 'tankless', 'tankless_coil'].includes(type) && efficiency_method === 'shipment_weighted') {
+        addErrorMessage(errorMessages[ENUMS.ERROR], `building/systems/domestic_hot_water/efficiency_method`, 'Invalid Efficiency Method for entered Hot Water Type');
+    }
+}
+
+function checkHotWaterYearValid(hot_water, errorMessages) {
+    const {year} = hot_water;
+    if(year && (1972 > year || (new Date()).getFullYear() < year)) {
+        addErrorMessage(errorMessages[ENUMS.BLOCKER], `building/systems/domestic_hot_water/year`, `Invalid year, must be between 1972 and ${(new Date()).getFullYear()}`)
+    }
+}
+
+function checkHotWaterEnergyFactorValid(hot_water, errorMessages) {
+    const {type, energy_factor} = hot_water;
+    if(["indirect", "tankless_coil"].includes(type) && !nullOrUndefined.includes(energy_factor)) {
+        addErrorMessage(errorMessages[ENUMS.BLOCKER], `building/systems/domestic_hot_water/energy_factor`, `Energy Factor not valid for selected Hot Water Type`);
+    }
+    let min,max;
+    if (type === 'storage') {
+        [min, max] = [0.45, 0.95];
+    } else if (type === 'tankless') {
+        [min, max] = [0.45, 0.99];
+    } else if (type === 'heat_pump') {
+        [min, max] = [1, 4];
+    }
+    if(energy_factor && (energy_factor < min || energy_factor > max)) {
+        addErrorMessage(errorMessages[ENUMS.BLOCKER], `building/systems/domestic_hot_water/energy_factor`, `${energy_factor} is outside the allowed range (${min} - ${max})`);
+    }
+}
+
+function checkSolarElectricYearValid(solar_electric, errorMessages) {
+    const {year} = solar_electric;
+    if(year && (2000 > year || (new Date()).getFullYear() < year)) {
+        addErrorMessage(errorMessages[ENUMS.BLOCKER], `building/systems/generation/solar_electric/year`, `Invalid year, must be between 2000 and ${(new Date()).getFullYear()}`)
+    }
+}
```

Comment:
It looks like this file is doing more than just getting the "required" fields, and handling the cross object validation as well.  I think it would be cleaner (and consistent with the previous implementation) for these to be in separate files.